### ISSUE TITLE
Missed optimization: long conditional jump on hot path

### ISSUE THREAD

**Issue by dvyukov** [llvm/llvm-project#80107]
Consider the following code:
```
int fast(int x, bool f) {
    if (__builtin_expect(f, false)) {
        return very_slow_path(x);
    }
    return x + 1;
}
```

Currently clang emits a long 6-byte conditional jump instruction on the hot path of the function:
```
 jne    1130 <_Z14very_slow_pathi>
```

I think for *very* hot functions it's better to emit a short 2-byte conditional jump on the hot path, and place long jump on a slow path. The nominal latency is the same for both, but more instruction bytes require more I cache space, and can move subsequent hot code to additional cache lines (?). This is hard to measure in micro-benchmarks.

https://godbolt.org/z/rqYf8cTE9

**Comment by dvyukov** [comment]
cc @david-xl

**Comment by david-xl** [comment]
yes, the branch folder can be taught to look at the branch probability and not fold the cold path tail call into the jcc in the hotpath.

**Comment by sesse** [comment]
A very similar problem:

```c++
size_t foo(size_t x)
{
    if (__builtin_expect(x < 200, true)) {
        return (x + 7) >> 3;
    } else {
        return (x + 127) >> 7;
    }
}
```

Clang will fold the tails of the two paths together, making the constant shifts into a variable-length shift (`shr rax, cl`) which is very slow on certain x86-64 CPUs.

**Comment by david-xl** [comment]
what I saw is worse -- Clang undid the folding, but fails to const-propagate  -- likely a phase ordering issue:

```asm
_Z3foom:
 mov    %rdi,%rax
 cmp    $0xc8,%rdi
 jae    117d <_Z3foom+0x1d>
 mov    $0x3,%ecx
 mov    $0x7,%edx
 add    %rdx,%rax
 shr    %cl,%rax
 ret
 mov    $0x7,%ecx
 mov    $0x7f,%edx
 add    %rdx,%rax
 shr    %cl,%rax
 ret
```

**Comment by sesse** [comment]
CE link to where it folds: https://gcc.godbolt.org/z/T1ffx8GPo

Of course, yours is even worse. :-)

**Comment by david-xl** [comment]
The difference is due to O2 vs O3.

**Comment by david-xl** [comment]
@weiguozhi

**Comment by weiguozhi** [comment]
The instruction

```asm
 jne    1130 <_Z14very_slow_pathi>
```

is actually a tail call, so it's difficult to call another function with an offset in the range [-128, 127).

**Comment by david-xl** [comment]
The idea is to jcc to another jmp instruction to save i-footprint on hotpath.

**Comment by weiguozhi** [comment]
> The idea is to jcc to another jmp instruction to save i-footprint on hotpath.

Then the idea is actually undo the conditional tail call. 
But the additional cold jmp instruction still need to be put into the instruction stream, within the offset range [-128, 127), not far away. Instruction cache line size is 64 bytes, and it usually fetch 2 blocks(128  bytes) at one time on modern x86. So my gut feeling is the benefit of this optimization is very small, and may hurt i-cache in some cases.
